File: connect_n.cpp
#include <cstdio>
#include <stdlib.h>
#include <random>
#include <chrono>
#include <assert.h>
#include <vector>
#include <algorithm>
using namespace std;

#define EMPTY 0
#define PLAYER1 1
#define PLAYER2 2

#define BOARD_SIZE 5 // Size of board
#define WIN_LENGTH 3 // Consecutive stones to win

// Computer will move 'O'
// Player will move 'X'
#define PLAYER1MOVE 'O'
#define PLAYER2MOVE 'X'

#define EMPTY_INIT 0
#define RANDOM_INIT 1

default_random_engine dre (chrono::steady_clock::now().time_since_epoch().count());     // provide seed
int random (int lim)
{
    uniform_int_distribution<int> uid {0,lim};   // help dre to generate nos from 0 to lim (lim included);
    return uid(dre);    // pass dre as an argument to uid to generate the random no
}

int random(int lim);
void printMove(int myMove);
int charToMove(char myChar, int myInt);
bool checkValidInput(char myChar, int myInt);
char playerToPlayerMove(int player);
int getHumanMove(int** board, int player);

class ConnectNBoard {
    private:
        int moves;
        int** board;
        int whichPlayerWin;
    
    public:
        int getMoves() {
            return moves;
        }

        int** getBoard() {
            return board;
        }

        int getWhichPlayerWin() {
            return whichPlayerWin;
        }

        ConnectNBoard(int flag) : moves(0), whichPlayerWin(0) {
            board = (int**) calloc(BOARD_SIZE, sizeof(int *));
            for (int i = 0; i < BOARD_SIZE; i++) {
                board[i] = (int *) calloc(BOARD_SIZE, sizeof(int));
            }
            if (flag == RANDOM_INIT) {
                int moves[3] = {EMPTY, PLAYER1, PLAYER2};
                for (int i = 0; i < BOARD_SIZE; i++) {
                    for (int j = 0; j < BOARD_SIZE; j++) {
                        board[i][j] = moves[random(2)];
                        if (board[i][j] != EMPTY) {
                            this->moves++;
                        }
                    }
                }    
            }
        }

        void showBoard() {
            printf("Board is showed below\n");
            char startChar = 'A';
            int startNum = BOARD_SIZE;
            printf("   ");
            for (int i = 0; i < BOARD_SIZE; i++) {
                printf(" %c  ", startChar++);
            }
            printf("\n");
            for (int i = 0; i < BOARD_SIZE; i++) {
                if (startNum < 10) {
                    printf(" ");
                }
                printf("%d ", startNum--);
                for (int j = 0; j < BOARD_SIZE; j++) {
                    printf(" %c ", playerToPlayerMove(board[i][j]));
                    if (j != BOARD_SIZE-1) {
                        printf("|");
                    }
                }
                printf("\n   ");
                for (int k = 0; k < BOARD_SIZE; k++) {
                    printf("----");
                }
                printf("\n");
            }
        }

        bool isWin(int player) {
            // Check horizontal
            int consec = 0;
            for (int i = 0; i < BOARD_SIZE; i++) {
                for (int j = 0; j < BOARD_SIZE; j++) {
                    if (board[i][j] == player) {
                        consec++;
                        if (consec == WIN_LENGTH) {
                            printf("player%d: Win by horizontal\n", player);
                            whichPlayerWin = player;
                            return true;
                        }
                    } else
                        consec = 0;
                }
                consec = 0;
            }

            // Check vertical
            consec = 0;
            for (int i = 0; i < BOARD_SIZE; i++) {
                for (int j = 0; j < BOARD_SIZE; j++) {
                    if (board[j][i] == player) {
                        consec++;
                        if (consec == WIN_LENGTH) {
                            printf("player%d: Win by vertical\n", player);
                            whichPlayerWin = player;
                            return true;
                        }
                    } else
                        consec = 0;
                }
                consec = 0;
            }

            // Check right diagonal
            consec = 0;
            int max_inspect;
            for (int i = 0; i < BOARD_SIZE; i++) {
                for (int j = 0; j < BOARD_SIZE; j++) {
                    max_inspect = BOARD_SIZE - max(i,j);
                    for (int k = 0; k < max_inspect; k++) {
                        if (board[i+k][j+k] == player) {
                            consec++;
                            if (consec == WIN_LENGTH) {
                                printf("player%d: Win by right diagonal\n", player);
                                whichPlayerWin = player;
                                return true;
                            }
                        }
                        else
                            consec = 0;
                    }
                    consec = 0;
                }
            }

            // Check left diagonal
            consec = 0;
            for (int i = 0; i < BOARD_SIZE; i++) {
                for (int j = 0; j < BOARD_SIZE; j++) {
                    max_inspect = min(BOARD_SIZE-i-1, j) + 1;
                    // printf("i:%d, j:%d, max_inspect:%d\n", i, j, max_inspect);
                    for (int k = 0; k < max_inspect; k++) {
                        if (board[i+k][j-k] == player) {
                            consec++;
                            if (consec == WIN_LENGTH) {
                                printf("player%d: Win by left diagonal\n", player);
                                whichPlayerWin = player;
                                return true;
                            }
                        }
                        else
                            consec = 0;
                    }
                    consec = 0;
                }
            }

            printf("player%d: Not win\n", player);
            return false;
        }

        int makeMove(int i, int j, int player) {
            int opponent;
            if (player == PLAYER1) {
                opponent = PLAYER2;
            } else if (player == PLAYER2) {
                opponent = PLAYER1;
            } else {
                printf("Wrong player\n");
                return -1;
            }

            if (moves >= BOARD_SIZE * BOARD_SIZE) {
                printf("The board is full\n");
                return -1;
            }
            if (i < 0 || i >= BOARD_SIZE
            || j < 0 || j >= BOARD_SIZE) {
                printf("Invalid move of i: %d, j: %d in makeMove\n", i, j);
                whichPlayerWin = opponent;
                return -1;
            }
            if (board[i][j] == EMPTY) {
                board[i][j] = player;
                moves++;
                return (i*BOARD_SIZE + j);
            } else {
                printf("Stone exists at i: %d, j: %d in makeMove\n", i, j);
                whichPlayerWin = opponent;
                return -1;
            }

        }

        int makeMove(int move, int player) {
            int opponent;
            if (player == PLAYER1) {
                opponent = PLAYER2;
            } else if (player == PLAYER2) {
                opponent = PLAYER1;
            } else {
                printf("Wrong player\n");
                return -1;
            }

            if (moves >= BOARD_SIZE * BOARD_SIZE) {
                printf("The board is full\n");
                return -1;
            }
            if (move < 0 || move >= BOARD_SIZE * BOARD_SIZE) {
                printf("Invalid move of %d in makeMove\n", move);
                whichPlayerWin = opponent;
                return -1;
            }
            int i = move / BOARD_SIZE;
            int j = move % BOARD_SIZE;
            if (board[i][j] == EMPTY) {
                board[i][j] = player;
                moves++;
                return (i*BOARD_SIZE + j);
            } else {
                printf("Stone exists at i: %d, j: %d in makeMove\n", i, j);
                whichPlayerWin = opponent;
                return -1;
            }
        }

        int makeRandomMove(int player) {
            int opponent;
            if (player == PLAYER1) {
                opponent = PLAYER2;
            } else if (player == PLAYER2) {
                opponent = PLAYER1;
            } else {
                printf("Wrong player\n");
                return -1;
            }
            if (moves >= BOARD_SIZE * BOARD_SIZE) {
                return -1;
            }
            int i,j;
            i = random(BOARD_SIZE-1);
            j = random(BOARD_SIZE-1);
            while (board[i][j] != EMPTY) {
                i = random(BOARD_SIZE-1);
                j = random(BOARD_SIZE-1);
            }
            board[i][j] = player;
            moves++;
            return i*BOARD_SIZE + j;
        }

        void playGame(int (*player1Func)(int**, int), int (*player2Func)(int**, int), bool showBoard) {
            bool gameDone = false;
            int whoseTurn = PLAYER1;
            int myMove;
            while (true) {
                if (moves >= BOARD_SIZE * BOARD_SIZE) {
                    break;
                }
                if (whoseTurn == PLAYER1) {
                    myMove = (player1Func)(board, PLAYER1);
                    myMove = this->makeMove(myMove, PLAYER1);
                    gameDone = this->isWin(PLAYER1);
                    if (myMove == -1)
                        break;
                    if (showBoard) {
                        this->showBoard();
                    }
                    if (gameDone)
                        break;
                    whoseTurn = PLAYER2;
                }
                else if (whoseTurn == PLAYER2) {
                    myMove = (player2Func)(board, PLAYER2);
                    myMove = this->makeMove(myMove, PLAYER2);
                    gameDone = this->isWin(PLAYER2);
                    if (myMove == -1)
                        break;
                    if (showBoard) {
                        this->showBoard();
                    }
                    if (gameDone)
                        break;
                    whoseTurn = PLAYER1;
                }
            }
        }

        int finishGame(bool printWinner) {
            if (printWinner) {
                if (whichPlayerWin == PLAYER1)
                    printf("Player1 won\n");
                else if (whichPlayerWin == PLAYER2)
                    printf("Player2 won\n");
                else
                    printf("No one won\n");
            }
            free(board);
            return whichPlayerWin;
        }
};

void printMove(int myMove) {
    char startChar = 'A';
    int startInt = BOARD_SIZE;
    startChar += myMove % BOARD_SIZE;
    startInt -= myMove / BOARD_SIZE;
    printf("%c%d", startChar, startInt);
}

int charToMove(char myChar, int myInt) {
    int offset = myChar - 'A';
    return (BOARD_SIZE - myInt) * BOARD_SIZE + offset;
}

bool checkValidInput(char myChar, int myInt) {
    int offset = myChar - 'A';
    if (offset < 0 || offset > BOARD_SIZE
    || myInt < 0 || myInt > BOARD_SIZE) {
        return false;
    }
    return true;
}

char playerToPlayerMove(int player) {
    if (player == EMPTY)
        return ' ';
    else if (player == PLAYER1)
        return PLAYER1MOVE;
    else if (player == PLAYER2)
        return PLAYER2MOVE;
    assert(0);
}

int getHumanMove(int** board, int player) {
    char myChar;
    int myInt;
    int myMove;
    int i,j;
    printf("player%d turn: ", player);
    while (true) {
        if (scanf("%c%d", &myChar, &myInt) == 2) {
            if (checkValidInput(myChar, myInt) == false) {
                printf("Invalid move, try again\n");
                continue;
            }
            myMove = charToMove(myChar, myInt);
            i = myMove / BOARD_SIZE;
            j = myMove % BOARD_SIZE;
            if (board[i][j] != EMPTY) {
                printf("Invalid move, try again\n");
                continue;
            }
            return myMove;
        }
    }
}


int main() {
    assert(BOARD_SIZE >= WIN_LENGTH);
    ConnectNBoard board(EMPTY_INIT);
    board.showBoard();
    board.playGame(getHumanMove, getHumanMove, true);

    board.finishGame(true);
}

File: connect_n.py
import numpy as np
from secrets import randbelow
import random

EMPTY = 0
PLAYER1 = 1
PLAYER2 = 2

BOARD_SIZE = 5
WIN_LENGTH = 3

PLAYER1MOVE = 'O'
PLAYER2MOVE = 'X'

EMPTY_INIT = 0
RANDOM_INIT = 1


class ConnectNBoard:
    def __init__(self, flag):
        self.moves = 0
        self.board = np.zeros(shape=(BOARD_SIZE, BOARD_SIZE))
        self.whichPlayerWin = 0
        if flag == RANDOM_INIT:
            moves = [EMPTY, PLAYER1, PLAYER2];
            for i in range(BOARD_SIZE):
                for j in range(BOARD_SIZE):
                    self.board[i][j] = moves[randbelow(3)]
                    if self.board[i][j] != EMPTY:
                        self.moves += 1

    def showBoard(self):
        startChar = ord('A')
        startNum = BOARD_SIZE
        print("   ", end='')
        for i in range(BOARD_SIZE):
            print(" %c  " % (chr(startChar)), end='')
            startChar += 1
        print()
        for i in range(BOARD_SIZE):
            if startNum < 10:
                print(" ", end='')
            print("%d " % startNum, end='')
            startNum -= 1
            for j in range(BOARD_SIZE):
                print(" %c " % playerToPlayerMove(self.board[i][j]), end='')  # change required
                if j != BOARD_SIZE - 1:
                    print("|", end='')
            print("\n   ", end='')
            for k in range(BOARD_SIZE):
                print("----", end='')
            print()

    def isWin(self, player):
        # Check horizontal
        consec = 0
        for i in range(BOARD_SIZE):
            for j in range(BOARD_SIZE):
                if self.board[i][j] == player:
                    consec += 1
                    if consec == WIN_LENGTH:
                        print("player%d: Win by horizontal" % player)
                        self.whichPlayerWin = player
                        return True
                else:
                    consec = 0
            consec = 0

        # Check vertical
        consec = 0
        for i in range(BOARD_SIZE):
            for j in range(BOARD_SIZE):
                if self.board[j][i] == player:
                    consec += 1
                    if consec == WIN_LENGTH:
                        print("player%d: Win by vertical" % player)
                        self.whichPlayerWin = player
                        return True
                else:
                    consec = 0
            consec = 0

        # Check right diagonal
        consec = 0
        for i in range(BOARD_SIZE):
            for j in range(BOARD_SIZE):
                max_inspect = BOARD_SIZE - max(i, j)
                for k in range(max_inspect):
                    if self.board[i+k][j+k] == player:
                        consec += 1
                        if consec == WIN_LENGTH:
                            print("player%d: Win by right diagonal" % player)
                            self.whichPlayerWin = player
                            return True
                    else:
                        consec = 0
                consec = 0

        # Check left diagonal
        consec = 0
        for i in range(BOARD_SIZE):
            for j in range(BOARD_SIZE):
                max_inspect = min(BOARD_SIZE-i-1, j) + 1
                for k in range(max_inspect):
                    if self.board[i-k][j-k] == player:
                        consec += 1
                        if consec == WIN_LENGTH:
                            print("player%d: Win by left diagonal" % player)
                            self.whichPlayerWin = player
                            return True
                    else:
                        consec = 0
                consec = 0

        print("player%d: Not win" % player)
        return False

    def makeMove(self, i, j, player):
        if player == PLAYER1:
            opponent = PLAYER2
        elif player == PLAYER2:
            opponent = PLAYER1
        else:
            print("Wrong player")
            return -1

        if self.moves >= BOARD_SIZE ** 2:
            print("The board is full")
            return -1
        if i < 0 or i > BOARD_SIZE or j < 0 or j >= BOARD_SIZE:
            print("Invalid move of i: %d, j: %d in makeMove" % (i, j))
            self.whichPlayerWin = opponent
            return -1
        if self.board[i][j] == EMPTY:
            self.board[i][j] = player
            self.moves += 1
            return i * BOARD_SIZE + j
        else:
            print("Stone exists at i: %d, j: %d in makeMove" % (i, j))
            self.whichPlayerWin = opponent
            return -1

    def makeMove(self, move, player):
        if player == PLAYER1:
            opponent = PLAYER2
        elif player == PLAYER2:
            opponent = PLAYER1
        else:
            print("Wrong player")
            return -1

        if self.moves >= BOARD_SIZE * BOARD_SIZE:
            print("The board is full")
            return -1
        if move < 0 or move >= BOARD_SIZE ** 2:
            print("Invalid move of %d in makeMove" % move)
            self.whichPlayerWin = opponent
            return -1
        i = int(move / BOARD_SIZE)
        j = move % BOARD_SIZE
        if self.board[i][j] == EMPTY:
            self.board[i][j] = player
            self.moves += 1
            return i * BOARD_SIZE + j
        else:
            print("Stone exists at i: %d, j: %d in makeMove" % (i, j))
            self.whichPlayerWin = opponent
            return -1

    def makeRandomMove(self, player):
        if player == PLAYER1:
            opponent = PLAYER2
        elif player == PLAYER2:
            opponent = PLAYER1
        else:
            print("Wrong player")
            return -1
        if self.moves >= BOARD_SIZE ** 2:
            return -1
        i = randbelow(BOARD_SIZE)
        j = randbelow(BOARD_SIZE)
        while self.board[i][j] != EMPTY:
            i = randbelow(BOARD_SIZE)
            j = randbelow(BOARD_SIZE)
        self.board[i][j] = player
        self.moves += 1
        return i * BOARD_SIZE + j

    def playGame(self, player1Func, player2Func, showBoard):
        gameDone = False
        whoseTurn = PLAYER1
        player1Param = (self.board, PLAYER1)
        player2Param = (self.board, PLAYER2)
        while True:
            if self.moves >= BOARD_SIZE ** 2:
                break
            if whoseTurn == PLAYER1:
                myMove = player1Func(*player1Param)
                myMove = self.makeMove(myMove, PLAYER1)
                gameDone = self.isWin(PLAYER1)
                if myMove == -1:
                    break
                if showBoard == True:
                    self.showBoard()
                if gameDone == True:
                    break
                whoseTurn = PLAYER2
            elif whoseTurn == PLAYER2:
                myMove = player2Func(*player2Param)
                myMove = self.makeMove(myMove, PLAYER2)
                gameDone = self.isWin(PLAYER2)
                if myMove == -1:
                    break
                if showBoard == True:
                    self.showBoard()
                if gameDone == True:
                    break
                whoseTurn = PLAYER1

    def finishGame(self, printWinner):
        if printWinner == True:
            if self.whichPlayerWin == PLAYER1:
                print("Player1 won")
            elif self.whichplayerWin == PLAYER2:
                print("Player2 won")
            else:
                print("No one won")
        return self.whichPlayerWin



def printMove(myMove):
    startChar = ord('A')
    startInt = BOARD_SIZE
    startChar += myMove % BOARD_SIZE
    startInt -= myMove / BOARD_SIZE
    print("%c%d" % (chr(startChar), startInt), end='')


def charToMove(myChar, myInt):
    offset = ord(myChar) - ord('A')
    print("myChar:" + myChar + " myInt:" + str(myInt))
    return (BOARD_SIZE - myInt) * BOARD_SIZE + offset


def checkValidInput(myChar, myInt):
    offset = ord(myChar) - ord('A')
    if offset < 0 or offset > BOARD_SIZE or myInt < 0 or myInt > BOARD_SIZE:
        return False
    return True


def playerToPlayerMove(player):
    if player == EMPTY:
        return ' '
    elif player == PLAYER1:
        return PLAYER1MOVE
    elif player == PLAYER2:
        return PLAYER2MOVE

def getHumanMove(board, player):
    while True:
        print("Player%d turn: " % player, end='')
        x = input()
        if len(x) < 2:
            print("Invalid move (too short), try again")
            continue
        myChar = x[0]
        myInt = x[1:]
        if not myInt.isdigit():
            print("Invalid move (not integer), try again")
            continue
        if ord(myChar) < ord('A') or ord(myChar) > ord('Z'):
            print("Invalid move (wrong character), try again")
            continue
        myInt = int(myInt)
        myMove = charToMove(myChar, myInt)
        i = int(myMove / BOARD_SIZE)
        j = myMove % BOARD_SIZE
        # print("i:" + str(i) + " and j:" + str(j))
        if board[i][j] != EMPTY:
            print("Invalid move, try again")
            continue
        return myMove



print("Hello from connect n")
gameboard = ConnectNBoard(EMPTY_INIT)
gameboard.showBoard()
gameboard.playGame(getHumanMove, getHumanMove, True)
gameboard.finishGame(True)
